services/supervisor/src/simple/mod.rs

```diff
@@ -22,17 +22,89 @@ use exonum::{
     runtime::{
         api::ServiceApiBuilder,
         rust::{CallContext, Service},
-        DispatcherError, ExecutionError, InstanceDescriptor, SUPERVISOR_INSTANCE_ID,
+        DispatcherError, ExecutionError, InstanceDescriptor, InstanceSpec, SUPERVISOR_INSTANCE_ID,
     },
 };
 use exonum_derive::{exonum_service, IntoExecutionError, ServiceFactory};
 
-use crate::{update_configs, ConfigChange, ConfigPropose, ConfigureCall};
+use crate::{
+    apply_consensus_config, apply_service_config, ConfigChange, ConfigPropose, ConfigureCall,
+};
 
 mod api;
 mod schema;
 pub use self::schema::Schema;
 
+/// Applies configuration changes, isolating each of them with by using `Fork` checkpoints.
+///
+/// # Safety
+///
+/// This function should be used with extreme care. It makes the following assumptions:
+///
+/// - The function must be called at the end of the transaction or `before_commit` execution.
+///   If the transaction errors / panics afterwards, the changes to the configs will not
+///   be rolled back.
+/// - No changes to the blockchain state should be introduced before the call to this function.
+///   Any changes that are introduced will be committed regardless of the execution status,
+///   or the status of application of any config change. This is if the execution wasn't interrupted
+///   by a panic / error *before* hitting the call; if this happens, the usual rules apply.
+///
+/// These restrictions are the result of `Fork` not having multi-layered checkpoints.
+fn update_configs(context: &mut CallContext, changes: Vec<ConfigChange>) {
+    // An error while configuring one of the service instances should not affect others.
+    changes.into_iter().for_each(|change| match change {
+        ConfigChange::Consensus(config) => {
+            log::trace!("Updating consensus configuration {:?}", config);
+            let configure_result = apply_consensus_config(context, config);
+
+            if let Err(e) = configure_result {
+                // Consensus configuration is not allowed to fail, since it was validated and approved.
+                // This is an unrecoverable error.
+                panic!("Consensus configuration failed with error {:?}", e)
+            }
+        }
+
+        ConfigChange::Service(config) => {
+            log::trace!(
+                "Updating service instance configuration, instance ID is {}",
+                config.instance_id
+            );
+            let configure_result = apply_service_config(context, config);
+
+            if let Err(e) = configure_result {
+                log::error!(
+                    "An error occurred while applying service configuration. {}",
+                    e
+                );
+            }
+        }
+
+        ConfigChange::DeployRequest(deploy_request) => {
+            log::trace!("Processing deploy request {:?}", &deploy_request);
+            // Activation height is ignored for simple supervisor.
```

Comment:
When `Supervisor` and `SimpleSupervisor` will use the same approach, we'll be able to remove `deadline_height` field.
For now I think ignoring the field value is better than implementing new almost-the-same entities.